lib/src/Pos/Network/Block/Retrieval.hs

```diff
@@ -341,16 +341,37 @@ streamProcessBlocks
     -> m ()
 streamProcessBlocks genesisConfig txpConfig diffusion nodeId desired checkpoints = do
     logInfo "streaming start"
-    r <- Diffusion.streamBlocks diffusion nodeId desired checkpoints writeCallback
+    mostDifficultBlock <- atomically $ newTVar Nothing
+    r <- Diffusion.streamBlocks diffusion nodeId desired checkpoints (writeCallback mostDifficultBlock)
     case r of
          Nothing -> do
              logInfo "streaming not supported, reverting to batch mode"
              getProcessBlocks genesisConfig txpConfig diffusion nodeId desired checkpoints
          Just _  -> do
              logInfo "streaming done"
-             return ()
+
+             recHeaderVar <- view (lensOf @RecoveryHeaderTag)
+             exitedRecovery <- atomically $ do
+                 mbMostDifficult <- readTVar mostDifficultBlock
+                 mbRecHeader <- tryReadTMVar recHeaderVar
+                 case (mbMostDifficult, mbRecHeader) of
+                      (Nothing, _)                 -> pure False -- We have not gotten a single block
+                      (Just _, Nothing)            -> pure False -- We where not in recovery?
+                      (Just mostDifficult, Just (_, recHeader)) ->
+                          if (mostDifficult ^. difficultyL) >= (recHeader ^. difficultyL)
+                             then isJust <$> tryTakeTMVar recHeaderVar
+                             else pure False
+
+             if exitedRecovery
+                then do
+                    logInfo "Recovery mode exited gracefully on receiving block we needed"
+                    return ()
+                else do -- Streaming stopped but we didn't make any progress
+                    _ <- dropRecoveryHeaderAndRepeat genesisConfig diffusion nodeId
+                    return ()
```

Comment:
This `if/then/else` block is suspect. `exitedRecovery` can be `False` if everything went fine (we just weren't in recovery), at which point we'll drop and repeat. Is that OK? Do we every stream blocks if we're _not_ in recovery mode?

Reply:
We will only stream blocks if we are in recovery mode, which can be set if we learn of that we are missing blocks or randomly as a failsafe. For normal operation we will fetch the block one at the time.